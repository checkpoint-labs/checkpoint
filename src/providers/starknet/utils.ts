import type { Abi } from 'starknet';

type StructAbi = {
  name: string;
  size: number;
  members: { name: string; type: string }[];
};

export const parseStruct = (
  type: string,
  data: string[],
  { current, structs }: { current: number; structs: Record<string, StructAbi> }
) => {
  const struct = structs[type];
  let structCurrent = current;

  return struct.members.reduce((output, field) => {
    if (structs[field.type]) {
      output[field.name] = parseStruct(field.type, data, { current: structCurrent, structs });
      structCurrent += structs[field.type].size;

      return output;
    }

    output[field.name] = data[structCurrent];
    structCurrent++;

    return output;
  }, {});
};

export const parseEvent = (abi: Abi, name: string, data: string[]): Record<string, any> => {
  const event = abi.find(el => el.name === name && el.type === 'event');
  if (!event) throw new Error('Unsupported event');

  const structs = abi
    .filter(el => el.type === 'struct')
    .reduce((acc, structAbi) => {
      acc[structAbi.name] = structAbi;
      return acc;
    }, {});

  let length = 0;
  let current = 0;

  return event.data.reduce((output, field) => {
    if (length > 0) {
      output[field.name] = data.slice(current, current + length);
      current += length;
      length = 0;

      return output;
    }

    if (structs[field.type]) {
      output[field.name] = parseStruct(field.type, data, { current, structs });
      current += structs[field.type].size;

      return output;
    }

    output[field.name] = data[current];
    if (field.name.endsWith('_len')) {
      length = parseInt(BigInt(data[current]).toString());
    }

    current++;

    return output;
  }, {});
};
