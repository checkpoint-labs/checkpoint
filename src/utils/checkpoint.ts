import { CheckpointConfig } from '../types';

export const getContractsFromConfig = (config: CheckpointConfig): string[] => {
  return config.sources.map(source => source.contract);
};
