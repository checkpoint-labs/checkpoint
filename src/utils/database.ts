import pluralize from 'pluralize';

export const getTableName = (name: string) => {
  if (name === '_metadata') return '_metadatas';

  return pluralize(name);
};
