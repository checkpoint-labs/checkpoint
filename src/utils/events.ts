type Callback<Item> = Item extends { name: infer Name; type: infer Type }
  ? Name extends PropertyKey
    ? Record<Name, Type>
    : never
  : never;

type Reducer<T extends Array<any>, Acc = Record<string, unknown>> = T extends []
  ? Acc
  : T extends [infer Head, ...infer Tail]
  ? Reducer<Tail, Acc & Callback<Head>>
  : never;

export const parseEvent = <Item extends { name: string; type: string }, Input extends Item[]>(
  format: readonly [...Input],
  input: string[]
): Reducer<Input> => {
  let length = 0;
  let current = 0;

  return format.reduce((output, field) => {
    if (length > 0) {
      output[field.name] = input.slice(current, current + length);
      current += length;
      length = 0;

      return output;
    }

    if (field.type === 'Uint256') {
      const uint256 = input.slice(current, current + 2);

      output[field.name] = (BigInt(uint256[0]) + (BigInt(uint256[1]) << BigInt(128))).toString();
      current += 2;

      return output;
    }

    output[field.name] = input[current];
    if (field.name.endsWith('_len')) {
      length = parseInt(BigInt(input[current]).toString());
    }

    current++;

    return output;
  }, {}) as Reducer<Input>;
};
