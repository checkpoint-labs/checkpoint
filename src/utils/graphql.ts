import {
  GraphQLObjectType,
  GraphQLOutputType,
  GraphQLNonNull,
  isLeafType,
  isListType,
  GraphQLScalarType
} from 'graphql';
import { jsonToGraphQLQuery } from 'json-to-graphql-query';
import pluralize from 'pluralize';

export const extendSchema = (schema: string): string => {
  return `directive @derivedFrom(field: String!) on FIELD_DEFINITION
${schema}`;
};

/**
 * Returns name of query for fetching single entity record
 *
 */
export const singleEntityQueryName = (entity: GraphQLObjectType) => entity.name.toLowerCase();

/**
 * Returns name of query for fetching multiple entity records
 *
 */
export const multiEntityQueryName = (entity: GraphQLObjectType) => {
  if (entity.name === '_Metadata') return '_metadatas';

  return pluralize(entity.name.toLowerCase());
};

/**
 * Generate sample query string based on entity object fields.
 *
 */
export const generateQueryForEntity = (entity: GraphQLObjectType): string => {
  // function to recursively build fields map
  const getObjectFields = (object: GraphQLObjectType, queryFields = {}): Record<string, any> => {
    const objectFields = object.getFields();

    Object.keys(objectFields).forEach(fieldName => {
      const rawFieldType = objectFields[fieldName].type;
      const fieldType = rawFieldType instanceof GraphQLNonNull ? rawFieldType.ofType : rawFieldType;

      if (isLeafType(fieldType)) {
        queryFields[fieldName] = true;
      } else if (isListType(fieldType)) {
        if (fieldType.ofType instanceof GraphQLScalarType) {
          queryFields[fieldName] = true;
        }
      } else {
        const childObjectFields = {};
        getObjectFields(fieldType as GraphQLObjectType, childObjectFields);
        queryFields[fieldName] = childObjectFields;
      }
    });

    return queryFields;
  };

  return jsonToGraphQLQuery(
    {
      query: {
        [multiEntityQueryName(entity)]: {
          __args: { first: 10 },
          ...getObjectFields(entity)
        }
      }
    },
    { pretty: true }
  );
};

export const getNonNullType = (type: GraphQLOutputType): GraphQLOutputType => {
  if (type instanceof GraphQLNonNull) {
    return type.ofType;
  }

  return type;
};
