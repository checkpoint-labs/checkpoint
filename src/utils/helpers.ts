export function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

export function chunk<T>(array: T[], chunkSize: number) {
  const chunks = [] as T[][];
  let index = 0;

  while (index < array.length) {
    chunks.push(array.slice(index, chunkSize + index));
    index += chunkSize;
  }

  return chunks;
}
