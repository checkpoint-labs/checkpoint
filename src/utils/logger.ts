import pino, { Logger as PinoLogger, LoggerOptions } from 'pino';

// LogLevel to control what levels of logs are
// required.
export enum LogLevel {
  // silent to disable all logging
  Silent = 'silent',
  // fatal to log unrecoverable errors
  Fatal = 'fatal',
  // error to log general errors
  Error = 'error',
  // warn to log alerts or notices
  Warn = 'warn',
  // info to log useful information
  Info = 'info',
  // debug to log debug and trace information
  Debug = 'debug'
}

type Logger = Omit<PinoLogger, 'trace'>;

export const createLogger = (opts: LoggerOptions = {}): Logger => {
  return pino(
    opts,
    pino.destination({
      sync: true
    })
  );
};

// re-export types as it is.
export { Logger, LoggerOptions };
