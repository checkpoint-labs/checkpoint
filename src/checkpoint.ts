import { GetBlockResponse, Provider } from 'starknet';
import { starknetKeccak } from 'starknet/utils/hash';
import { validateAndParseAddress } from 'starknet/utils/address';
import Promise from 'bluebird';
import getGraphQL, { CheckpointsGraphQLObject, MetadataGraphQLObject } from './graphql';
import { GqlEntityController } from './graphql/controller';
import { createLogger, Logger, LogLevel } from './utils/logger';
import { AsyncMySqlPool, createMySqlPool } from './mysql';
import { CheckpointConfig, CheckpointOptions, SupportedNetworkName } from './types';
import { getContractsFromConfig } from './utils/checkpoint';
import { CheckpointRecord, CheckpointsStore, MetadataId } from './stores/checkpoints';
import { GraphQLObjectType } from 'graphql';

export default class Checkpoint {
  public config: CheckpointConfig;
  public writer;
  public schema: string;
  public provider: Provider;

  private readonly entityController: GqlEntityController;
  private readonly log: Logger;
  private readonly sourceContracts: string[];

  private mysqlPool?: AsyncMySqlPool;
  private mysqlConnection?: string;
  private checkpointsStore?: CheckpointsStore;
  private cpBlocksCache: number[] | null;

  constructor(config: CheckpointConfig, writer, schema: string, opts?: CheckpointOptions) {
    this.config = config;
    this.writer = writer;
    this.schema = schema;
    this.entityController = new GqlEntityController(schema);
    this.provider = new Provider({ network: this.config.network as SupportedNetworkName });
    this.sourceContracts = getContractsFromConfig(config);
    this.cpBlocksCache = [];

    this.log = createLogger({
      base: { component: 'checkpoint' },
      level: opts?.logLevel || LogLevel.Error,
      ...(opts?.prettifyLogs
        ? {
            transport: {
              target: 'pino-pretty'
            }
          }
        : {})
    });

    this.mysqlConnection = opts?.dbConnection;
  }

  /**
   * Returns an express handler that exposes a GraphQL API to query entities defined
   * in the schema.
   *
   */
  public get graphql() {
    const entityQueryFields = this.entityController.generateQueryFields();
    const coreQueryFields = this.entityController.generateQueryFields([
      MetadataGraphQLObject,
      CheckpointsGraphQLObject
    ]);

    const querySchema = new GraphQLObjectType({
      name: 'Query',
      fields: {
        ...entityQueryFields,
        ...coreQueryFields
      }
    });

    return getGraphQL(querySchema, {
      log: this.log.child({ component: 'resolver' }),
      mysql: this.mysql
    });
  }

  /**
   * Starts the indexer.
   *
   * The indexer will invoker the respective writer functions when a contract
   * event is found.
   *
   */
  public async start() {
    this.log.debug('starting');

    const blockNum = await this.getStartBlockNum();
    return await this.next(blockNum);
  }

  /**
   * Reset will clear the last synced block informations
   * and force Checkpoint to start indexing from the start
   * block.
   *
   * This will also clear all indexed GraphQL entity records.
   *
   * This should be called when there has been a change to the GraphQL schema
   * or a change to the writer functions logic, so indexing will re-run from
   * the starting block. Also, it should be called the first time Checkpoint
   * is being initialized.
   *
   */
  public async reset() {
    this.log.debug('reset');

    await this.store.createStore();
    await this.store.setMetadata(MetadataId.LastIndexedBlock, 0);

    await this.entityController.createEntityStores(this.mysql);
  }

  /**
   * Registers the blocks where a contracts event can be found.
   * This will be used as a skip list for checkpoints while
   * indexing relevant blocks. Using this seed function can significantly
   * reduce the time for Checkpoint to re-index blocks.
   *
   * This should be called before the start() method is called.
   *
   */
  public async seedCheckpoints(
    checkpointBlocks: { contract: string; blocks: number[] }[]
  ): Promise<void> {
    await this.store.createStore();

    const checkpoints: CheckpointRecord[] = [];

    checkpointBlocks.forEach(cp => {
      cp.blocks.forEach(blockNumber => {
        checkpoints.push({ blockNumber, contractAddress: cp.contract });
      });
    });

    await this.store.insertCheckpoints(checkpoints);
  }

  private async getStartBlockNum() {
    let start = 0;
    let lastBlock = await this.store.getMetadataNumber(MetadataId.LastIndexedBlock);
    lastBlock = lastBlock ?? 0;

    const nextBlock = lastBlock + 1;

    this.config.sources.forEach(source => {
      start = start === 0 || start > source.start ? source.start : start;
    });
    return nextBlock > start ? nextBlock : start;
  }

  private async getNextBlockNumber(blockNumber: number) {}

  private async next(blockNum: number) {
    const checkpointBlock = await this.getNextCheckpointBlock(blockNum);
    if (checkpointBlock) {
      blockNum = checkpointBlock;
    }

    this.log.debug({ blockNumber: blockNum }, 'next block');

    let block: GetBlockResponse;
    try {
      block = await this.provider.getBlock(blockNum);
    } catch (e) {
      this.log.error({ blockNumber: blockNum, err: e }, 'getting block failed... retrying');

      await Promise.delay(12e3);
      return this.next(blockNum);
    }

    await this.handleBlock(block);

    await this.store.setMetadata(MetadataId.LastIndexedBlock, block.block_number);

    return this.next(blockNum + 1);
  }

  private async getNextCheckpointBlock(blockNum: number): Promise<number | null> {
    if (this.cpBlocksCache === null) {
      // cache is null when we can no more find a record in the database
      // so exiting early here to avoid polling the database in subsequent
      // loops.
      return null;
    }

    if (this.cpBlocksCache.length !== 0) {
      return this.cpBlocksCache.shift();
    }

    const checkpointBlocks = await this.store.getNextCheckpointBlocks(
      blockNum,
      this.sourceContracts
    );
    if (checkpointBlocks.length === 0) {
      this.log.info({ blockNumber: blockNum }, 'no more checkpoint blocks in store');
      // disabling cache to stop polling database
      this.cpBlocksCache = null;

      return null;
    }

    this.cpBlocksCache = checkpointBlocks;
    return this.cpBlocksCache.shift();
  }

  private async handleBlock(block: GetBlockResponse) {
    this.log.info({ blockNumber: block.block_number }, 'handling block');

    // @ts-ignore
    for (const receipt of block.transaction_receipts) {
      await this.handleTx(block, block.transactions[receipt.transaction_index], receipt);
    }

    this.log.debug({ blockNumber: block.block_number }, 'handling block done');
  }

  private async handleTx(block, tx, receipt) {
    this.log.debug({ txIndex: tx.transaction_index }, 'handling transaction');

    for (const source of this.config.sources) {
      let foundContractData = false;
      const contract = validateAndParseAddress(source.contract);

      if (contract === validateAndParseAddress(tx.contract_address)) {
        if (tx.type === 'DEPLOY' && source.deploy_fn) {
          foundContractData = true;
          this.log.info(
            { contract: source.contract, txType: tx.type, handlerFn: source.deploy_fn },
            'found deployment transaction'
          );

          await this.writer[source.deploy_fn]({ source, block, tx, receipt, mysql: this.mysql });
        }
      }

      for (const event of receipt.events) {
        if (contract === validateAndParseAddress(event.from_address)) {
          for (const sourceEvent of source.events) {
            if (`0x${starknetKeccak(sourceEvent.name).toString('hex')}` === event.keys[0]) {
              foundContractData = true;
              this.log.info(
                { contract: source.contract, event: sourceEvent.name, handlerFn: sourceEvent.fn },
                'found contract event'
              );

              await this.writer[sourceEvent.fn]({ source, block, tx, receipt, mysql: this.mysql });
            }
          }
        }
      }

      if (foundContractData) {
        await this.store.insertCheckpoints([
          { blockNumber: block.block_number, contractAddress: source.contract }
        ]);
      }
    }

    this.log.debug({ txIndex: tx.transaction_index }, 'handling transaction done');
  }

  private get store(): CheckpointsStore {
    if (this.checkpointsStore) {
      return this.checkpointsStore;
    }

    return (this.checkpointsStore = new CheckpointsStore(this.mysql, this.log));
  }

  private get mysql(): AsyncMySqlPool {
    if (this.mysqlPool) {
      return this.mysqlPool;
    }

    // lazy initialization of mysql connection
    return (this.mysqlPool = createMySqlPool(this.mysqlConnection));
  }
}
