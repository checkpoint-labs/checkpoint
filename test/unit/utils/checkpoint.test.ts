import { getContractsFromConfig } from '../../../src/utils/checkpoint';
import { validCheckpointConfig } from '../../fixtures/checkpointConfig.fixture';

describe('getContractsFromConfig', () => {
  it('should work', () => {
    expect(getContractsFromConfig(validCheckpointConfig)).toMatchSnapshot();
  });
});
