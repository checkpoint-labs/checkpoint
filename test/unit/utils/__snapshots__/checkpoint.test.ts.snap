// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getContractsFromConfig should work 1`] = `
Array [
  "0x0625dc1290b6e936be5f1a3e963cf629326b1f4dfd5a56738dea98e1ad31b7f3",
]
`;
